include/event: Handle error code returns from event handlers

jschlumpp opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

According to the comments the handlers should be able to return error codes to signal a failure. This fixes the event handler loop to handle these negative return values. We stop the handler iteration when a handler returns an error.

<!--
Please provide a detailed description of the changes made in this new PR.
-->

jschlumpp committed:
include/event: Handle error code returns from event handlers

According to the comments the handlers should be able to return error
codes to signal a failure. This fixes the event handler loop to handle
these negative return values. We stop the handler iteration when a
handler returns an error.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

marcrittinghaus commented on include/uk/event.h:
```diff
@@ -267,8 +267,8 @@ static inline int uk_raise_event_ptr(struct uk_event *e, void *data)
 		__uk_event_assert(*itr);
-		ret |= ((*itr)(data));
-		if (ret == UK_EVENT_HANDLED)
+		ret = ((*itr)(data));
+		if (ret < 0 || ret == UK_EVENT_HANDLED)
```
A successful return must be sticky, which is not the case in your solution. E.g., you could have `HANDLED_CONT`->`NOT_HANDLED` and it would return `NOT_HANDLED` but should return `HANDLED(_CONT)`.

The code should look like this:
```C
		...
		rc = ((*itr)(data));
		if (rc < 0 || rc == UK_EVENT_HANDLED)
			return rc;
		ret |= rc;
	}
```

jschlumpp replied:
Good point. However, were the enum values intended to be able to represent a bit-mask? In your proposed change the `raise_event` could also return `UK_EVENT_HANDLED_CONT | UK_EVENT_HANDLED == 3`.

marcrittinghaus replied:
Right, I also had a variant with this instead of the OR.
```C
if (rc == UK_EVENT_HANDLED_CONT)
   ret = UK_EVENT_HANDLED;
```
It is just an additional branch that I wanted to avoid. But you are right, taking the function description into account the OR variant is not ok.

marcrittinghaus reviewed (changes requested):
Hey @mschlumpp,

just wanted to test this while debugging another issue and noticed it does not work. Please see my comment.

Another issue very closely related to this is that the calling sites of `uk_raise_event(_ptr)()` only check for a non-zero value to accept it as handled. But since the function might also return an `errno` the fault should be considered not handled for negative return values. Means only `> 0` means success.

Could you fix these as part of the PR?

jschlumpp committed:
plat/common: Adjust uk_raise_event call location

The event system is currently used for traps. In case, an event handler
returns an error, this prints an additional message and then continues with
the normal unhandled-trap case.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
include/event: Fix parameter name in documentation

The parameter name in the documentation comment did not match the actual
parameter name of the function definition.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

craciunoiuc reviewed (commented):
Hey @mschlumpp 

I'm not 100% sure how I should test this, but I tried some crashes and I guess things worked.

gabrielmocanu commented on include/uk/event.h:
```diff
@@ -251,25 +251,30 @@ struct uk_event {
+		rc = ((*itr)(data));
+		if (rc < 0 || rc == UK_EVENT_HANDLED)
 			break;
+		if (rc == UK_EVENT_HANDLED)
```
Is this correct? For me, it seems that this code won't be executed.

jschlumpp replied:
Oops. You are right. Good catch.

jschlumpp commented:
@craciunoiuc You could write some event handlers returning the various return values to check whether it works.

craciunoiuc commented:
> @craciunoiuc You could write some event handlers returning the various return values to check whether it works.

I feel stupid, I can't seem to get it to work. The [code comments](https://github.com/unikraft/unikraft/pull/691/files#diff-ef64eda0b3f99a203d7e108026575ac468a50a498579f2881cbb5a07c932e267L106)  (around Line 106 in `include/uk/event.h`) might be out of date.

With the comments there I think I created an event successfully, but I can't register the handlers, cause the call is incorrect (missing 1 argument).

Edit: Spoke too soon, got it to work, but the comment example is still wrong :D

jschlumpp commented:
Yes, the documentation seems to be wrong.
The macro is defined as:
```c
UK_EVENT_HANDLER_PRIO(event, fn, prio)
```
but the documentation uses it as:
```c
UK_EVENT_HANDLER_PRIO(handler1, UK_PRIO_EARLIEST);
```

This is what you are talking about, right?

jschlumpp committed:
include/event: Fix usage of UK_EVENT_HANDLER_PRIO in example

The example missed the event parameter, which must be specified.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

craciunoiuc commented:
Yup, I was testing it now, almost done, it works btw ;)

craciunoiuc commented on include/uk/event.h:
```diff
@@ -105,9 +105,9 @@
- * UK_EVENT_HANDLER_PRIO(handler1, UK_PRIO_EARLIEST);
- * UK_EVENT_HANDLER_PRIO(handler2, UK_PRIO_LATEST);
- * UK_EVENT_HANDLER_PRIO(handler3, UK_PRIO_LATEST);
+ * UK_EVENT_HANDLER_PRIO(event, handler1, UK_PRIO_EARLIEST);
+ * UK_EVENT_HANDLER_PRIO(event, handler2, UK_PRIO_LATEST);
+ * UK_EVENT_HANDLER_PRIO(event, handler3, UK_PRIO_LATEST);
```
Might as well go ahead and also mention in the comments how to trigger an event, I think this is also missing

```c
        struct myevent_data data;
        // ...
        int ret = uk_raise_event(myevent, &data);
```

jschlumpp replied:
That's already there https://github.com/unikraft/unikraft/blob/ac15f671186430848629313836715616687ff4e5/include/uk/event.h#L79

craciunoiuc replied:
whoops, missed it, closing this

craciunoiuc reviewed (changes requested):
@mschlumpp 

Only one last comment from my side.

Also, check the codacy errors, maybe they are fixable.

marcrittinghaus commented:
Thanks guys for fixing the glitches in the doc and the code. Crazy how that slipped through although I can remember that we had a fairly long discussion about the events.

marcrittinghaus commented on include/uk/event.h:
```diff
@@ -251,24 +253,32 @@ struct uk_event {
-		ret |= ((*itr)(data));
-		if (ret == UK_EVENT_HANDLED)
+		rc = ((*itr)(data));
+		if (rc < 0)
+			return rc;
```
I won't get tired saying this. Can you please add the `unlikely` for error cases 😛

craciunoiuc reviewed (commented):
All good here. Thanks!

I'll wait with the review tag for @marcrittinghaus

marcrittinghaus reviewed (approved):
Hi @mschlumpp,

thanks 😃 Looks good!

Approved-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
